### ISSUE TITLE
Hap Alpha video not working on AMD graphics card

### ISSUE THREAD

**Issue by tobiasebsen** [bangnoise/ofxHapPlayer#50]
I have a strange issue.
My application plays some Hap Alpha files. Everything works fine on two different Windows computers (one has NVIDIA Titan card, the other has an integrated Intel Iris Plus 650). It even works on my Mac with a NVIDIA GT 750M card. But...

When I try the application on a machine with a AMD Radeon RX Vega – there is no image drawn. I have also tried it with another (identical) computer with no luck.

Then I tried to swap the files to normal Hap (no alpha) – and the video shows perfectly through the AMD graphics card. Strange.

So... something about Hap Alpha compressed textures seems to not be compatible with the AMD drivers. Has anyone had similar issues?

BTW. I'm on WIndows 10. Tried both 32-bit and 64-bit binaries. No difference.

**Comment by tobiasebsen** [comment]
A bit more investigation reveals that glCompressedTexSubImage2D fails with GL_INVALID_OPERATION.

I have also checked the list of supported compression formats using glGet GL_COMPRESSED_TEXTURE_FORMATS and it does indeed include GL_COMPRESSED_RGBA_S3TC_DXT5_EXT

**Comment by bangnoise** [comment]
Strange - AMD drivers up to date? Does this also happen with the ofxHapPlayer example app on the affected machine?

**Comment by tobiasebsen** [comment]
Yes. It's a new computer with most recent drivers installed. And the issues is the same with the example app.

**Comment by sunekastbjerg** [comment]
I'm having the same strange issue. The issue only happens on AMD cards (i tried an RX 480 and R9 290), and it has not been an issue with the last versions of ofxHapPlayer. I have found that it is fixed by making the video dimensions divisible by 4. I had a 985 x 574 video that works fine on AMD cards if i scale it to 984 x 572. If i just make the dimensions even, it does not work.

**Comment by bangnoise** [comment]
@sunekastbjerg ah - OK, helpful to know it works for multiples of 4. Could you try the following change to ofxHapPlayer.cpp, lines 546 onwards:

```
#if OFX_HAP_HAS_CODECPAR
            ofxHapPY::roundUpToMultipleOf4(_videoStream->codecpar->width),
            ofxHapPY::roundUpToMultipleOf4(_videoStream->codecpar->height),
#else
            ofxHapPY::roundUpToMultipleOf4(_videoStream->codec->width),
            ofxHapPY::roundUpToMultipleOf4(_videoStream->codec->height),
#endif
```

Then try playing an odd-dimensioned movie and report back. Thanks!

**Comment by bangnoise** [comment]
just for clarity, the full function call there becomes

````
        glCompressedTexSubImage2D(GL_TEXTURE_2D,
            0,
            0,
            0,
#if OFX_HAP_HAS_CODECPAR
            ofxHapPY::roundUpToMultipleOf4(_videoStream->codecpar->width),
            ofxHapPY::roundUpToMultipleOf4(_videoStream->codecpar->height),
#else
            ofxHapPY::roundUpToMultipleOf4(_videoStream->codec->width),
            ofxHapPY::roundUpToMultipleOf4(_videoStream->codec->height),
#endif
            internalFormat,
            static_cast<GLsizei>(_decodedFrame.buffer.size()),
            _decodedFrame.buffer.data());
````

**Comment by sunekastbjerg** [comment]
Thanks for the fast response! I'm getting an GL_INVALID_VALUE from glCompressedTexSubImage2D with those values. Don't we need to resize the buffer as well for that to work?

**Comment by bangnoise** [comment]
The buffer is correctly sized because DXT data is always in 4x4 blocks even if some of the pixels are beyond the bounds of the image (drivers are expected to be able to handle odd widths in glCompressedTexSubImage2D, so this is a driver bug).

We probably need to allocate the texture with the same rounded dimensions - but we don't want OF to start drawing the full texture either. See if it works if you perform the same rounding at lines 513 onwards where we set the texture's dimensions - it's not a solution because you'll also see the extra pixels when you draw the texture, but we can force that if we discover this works.

**Comment by tobiasebsen** [comment]
There is also the ``coded_width`` and ``coded_height`` parameters to reveal the actual dimensions of the data. They should be the same as multiple of four, but coded dimensions are more correct, right?

**Comment by bangnoise** [comment]
@tobiasebsen from `AVCodecContext`? We don't have one for the video stream.

**Comment by sunekastbjerg** [comment]
> The buffer is correctly sized because DXT data is always in 4x4 blocks even if some of the pixels are beyond the bounds of the image (drivers are expected to be able to handle odd widths in glCompressedTexSubImage2D, so this is a driver bug).
> 
> We probably need to allocate the texture with the same rounded dimensions - but we don't want OF to start drawing the full texture either. See if it works if you perform the same rounding at lines 513 onwards where we set the texture's dimensions - it's not a solution because you'll also see the extra pixels when you draw the texture, but we can force that if we discover this works.

Just tried that out, and it seems to work with the odd sized textures when the allocation is rounded up to multiples of 4 as well. Thanks! :)

**Comment by tobiasebsen** [comment]
@bangnoise you could just set the ``_texture.texData.width`` and ``_texture.texData.height`` to the non-rounded values after texture allocation to avoid drawing the extra pixels.

**Comment by tobiasebsen** [comment]
... and, of cause, scale the ``tex_t`` and ``tex_u`` accordingly.

**Comment by tobiasebsen** [comment]
#63 Like this ?

**Comment by sunekastbjerg** [comment]
Yes that is what i tried and it works on an AMD R9 290 at least. :)

**Comment by bangnoise** [comment]
Fixed in #63 - thanks @tobiasebsen
